src/lib.rs

```diff
@@ -5187,13 +5187,64 @@ where
     pub fn into_ref(self) -> &'a T {
         // Presumably unreachable, since we've guarded each constructor of `Ref`.
         util::assert_dst_is_not_zst::<T>();
-        // PANICS: By invariant on `Ref`, `self.0`'s size and alignment are
-        // valid for `T`, and so this `unwrap` will not panic.
-        let ptr = Ptr::from_ref(self.0.into())
-            .try_cast_into_no_leftover::<T>()
-            .expect("zerocopy internal error: into_ref should be infallible");
-        let ptr = ptr.bikeshed_recall_valid();
-        ptr.as_ref()
+        match Ptr::from_ref(self.0.into()).try_cast_into_no_leftover::<T>() {
+            Ok(ptr) => ptr.bikeshed_recall_valid().as_ref(),
+            Err(CastError::Validity(i)) => match i {},
+            Err(CastError::Alignment(_) | CastError::Size(_)) => {
+                // SAFETY: By invariant on `Ref::0`, the referenced byte slice
+                // is aligned to `T`'s alignment and its size corresponds to a
+                // valid size for `T`. Since properties are checked upon
+                // constructing `Ref`, these failures are unreachable.
+                unsafe { core::hint::unreachable_unchecked() }
+            }
+        }
+    }
+
+    /// Attempts to convert this `Ref<_, T>` into a `&T`.
+    ///
+    /// `try_into_ref` consumes the `Ref`, and returns a reference to `T`.
+    #[must_use = "has no side effects"]
+    #[inline(always)]
+    pub fn try_into_ref(self) -> Result<&'a T, ValidityError<Self, T>>
+    where
+        T: TryFromBytes,
+    {
+        util::assert_dst_is_not_zst::<T>();
+        let is_bit_valid = match Ptr::from_ref(&*self.0).try_cast_into_no_leftover::<T>() {
+            Ok(candidate) => TryFromBytes::is_bit_valid(candidate.forget_aligned()),
+            Err(CastError::Validity(i)) => match i {},
+            Err(CastError::Alignment(_) | CastError::Size(_)) => {
+                // SAFETY: By invariant on `Ref::0`, the referenced byte slice
+                // is aligned to `T`'s alignment and its size corresponds to a
+                // valid size for `T`. Since properties are checked upon
+                // constructing `Ref`, these failures are unreachable.
+                unsafe { core::hint::unreachable_unchecked() }
+            }
+        };
+        if !is_bit_valid {
+            Err(ValidityError::new(self))
+        } else {
+            match Ptr::from_ref(self.0.into()).try_cast_into_no_leftover::<T>() {
```

Comment:
Is there any way we can keep the original `Ptr` and avoid calling constructing it twice?

Reply:
Not that I can find, unfortunately. It might be possible with more `unsafe` code.

Reply:
We might be able to do something like this, but lifetime extension makes me nervous:
```rust
    /// Attempts to convert this `Ref<_, T>` into a `&mut T`.
    ///
    /// `try_into_mut` consumes the `Ref`, and returns a reference to `T`.
    #[must_use = "has no side effects"]
    #[inline(always)]
    pub fn try_into_mut(mut self) -> Result<&'a mut T, ValidityError<Self, T>>
    where
        T: TryFromBytes,
    {
        // Extend the lifetime of `&mut *self.0` to `'a`.
        // SAFETY: Idk, this feels sketchy.
        let bytes = unsafe { mem::transmute::<&mut [u8], &'a mut [u8]>(&mut *self.0) };
        match Ptr::from_mut(bytes).try_cast_into_no_leftover::<T>() {
            Ok(candidate) => {
                match candidate.try_into_valid() {
                    Ok(candidate) => Ok(candidate.as_mut()),
                    Err(e) => Err(e.with_src(self)),
                }
            }
            Err(CastError::Validity(i)) => match i {},
            Err(CastError::Alignment(_) | CastError::Size(_)) => {
                // SAFETY: By invariant on `Ref::0`, the referenced byte slice
                // is aligned to `T`'s alignment and its size corresponds to a
                // valid size for `T`. Since properties are checked upon
                // constructing `Ref`, these failures are unreachable.
                unsafe { core::hint::unreachable_unchecked() }
            },
        }
    }
```

Reply:
If `IntoByteSlice` documented that `.into()` behaved identically to `.deref()`, this might be fine?